core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -576,13 +587,113 @@ private static boolean representAsInternalType(Type type) {
    * Handle decimal type specifically with explicit type conversion
    */
   private static Expression convertAssignableType(
-      Expression argument,  Type targetType) {
+      Expression argument, Type targetType) {
     if (targetType != BigDecimal.class) {
       return argument;
     }
     return convert(argument, targetType);
   }
 
+  public static MethodCallExpression call(Class clazz, String methodName,
+      List<? extends Expression> arguments) {
+    return call(clazz, methodName, arguments, null);
+  }
+
+  /**
+   * The powerful version of {@code org.apache.calcite.linq4j.tree.Expressions#call(
+   * Type, String, Iterable<? extends Expression>)}. Try best effort to convert the
+   * accepted arguments to match parameter type.
```

Comment:
Is there a reason the method is put into `core` module rather than into `linq4j`?

Reply:
You mean this file ?

Reply:
I mean that the added `call` method looks to belong to `linq4j` more that to `core`.
The comment explicitly says this is a "more powerful" method, so it suggests that either we replace the existing linq4j method or add an additional one side by side, so everybody could compare and select the appropriate one.

Having "similar yet slightly different methods" in different modules/classes is sad.

Reply:
Agree with @vlsi .

Reply:
@vlsi @hsyuan The main reason is type cast function.
In [CALCITE-3414](https://issues.apache.org/jira/browse/CALCITE-3414), we ever made a discussion on unifying the type cast/convert function.

> Expressions.convert and Types.castIfNecessary follow that policy. RexToLixTranslator.convert has a mixture - some stuff regarding boxing/unboxing, and primitive to primitive (e.g. float to double), that should be in linq4j, and some stuff such as converting java.sql.Timestamp to long and BigDecimal to String that would remain in core.

Therefore, we unify the implementations in `EnumUtils.convert` and put it in calcite core. 

The `call` method relies on `EnumUtils.convert`, especially for Decimal types, which cannot be handled in `linq4j`.

Reply:
Can we define an interface for argument type check/convert. `Linq4j` adopts a default one, c`alcite-core` develops a more powerful one?

Reply:
An interface (`ArgumentTypeMatcher`) involving type check and match is defined. Users/developers can pass customized implementions to construct a `MethodCallExpression`.

Reply:
I don't think we should add any interface for code generation, which is already Calcite implementation. Put it into code module is okey for me, just like Julian said, linq4j follows type assignment/style of Java, we do not need to follow that here.

Reply:
Withou breaking the policy, the interface is defined internally in `Types` class in `Linq4j`. It is just used to differentiate beheaviors. The default one in `Linq4j` is implemented with type assignment/style of Java, while the one in `calcite-core` extends functionalities.
Another potential benefit:Many component (Cassandra, ElasticSearch, etc.) also lookup methods based on linq4j. 

But , of course, I can change it back. Both of them are ok to me.

Reply:
Please change it back, i didn't see any necessity to do that at least for now which would bring in complexity without obvious benefit.

Reply:
Thanks for direction, changed back and rebased.